### PR TITLE
Fix typo annotations in JSDoc with TypeScript

### PR DESCRIPTION
**Pull request by geovanisouza92**
Hi, thanks for this amazing lib!

This PR enable TypeScript to enhance type checking and experience for TS users (me included).

There are some small typos that I think can cause bugs here and there, I'll flag them on comments.

---

I think that migrate to TS can enhance even more the quality of the lib, and if you agree, I can make another PR later.

### CODE REVIEW

File: lib/Deque.js [hunk 1]
```diff
@@ -13,10 +13,10 @@ class Deque {
 
   /**
    * removes and returns the first element from the queue
-   * @return {[type]} [description]
+   * @return {any} [description]
    */
   shift() {
-    if (this._length === 0) {
+    if (this.length === 0) {
```
Comment (geovanisouza92) [inline lib/Deque.js]: Here, `this._length` points to nothing, while other methods use `this.length` that is a getter for `this._list.length`

File: lib/Deque.js [hunk 2]
```diff
@@ -52,7 +52,7 @@ class Deque {
    * removes and returns the last element from the queue
    */
   pop() {
-    if (this._length === 0) {
+    if (this.length === 0) {
```
Comment (geovanisouza92) [inline lib/Deque.js]: Same thing above

File: lib/Deque.js [hunk 3]
```diff
@@ -76,10 +76,10 @@ class Deque {
 
   /**
    * get a reference to the item at the head of the queue
-   * @return {element} [description]
+   * @return {any} [description]
    */
   get head() {
-    if (this._list.length === 0) {
+    if (this.length === 0) {
```
Comment (geovanisouza92) [inline lib/Deque.js]: Same thing above (more like standarization)

File: lib/Deque.js [hunk 4]
```diff
@@ -88,10 +88,10 @@ class Deque {
 
   /**
    * get a reference to the item at the tail of the queue
-   * @return {element} [description]
+   * @return {any} [description]
    */
   get tail() {
-    if (this._list.length === 0) {
+    if (this.length === 0) {
```
Comment (geovanisouza92) [inline lib/Deque.js]: Same thing above

File: lib/Pool.js [hunk 5]
```diff
@@ -66,7 +74,7 @@ class Pool extends EventEmitter {
     /**
      * A queue/stack of pooledResources awaiting acquisition
      * TODO: replace with LinkedList backed array
-     * @type {Array}
+     * @type {Deque}
```
Comment (geovanisouza92) [inline lib/Pool.js]: Outdated type

File: lib/Pool.js [hunk 6]
```diff
@@ -90,7 +98,7 @@ class Pool extends EventEmitter {
 
     /**
      * All objects associated with this pool in any state (except destroyed)
-     * @type {PooledResourceCollection}
+     * @type {Set}
```
Comment (geovanisouza92) [inline lib/Pool.js]: Outdated type

File: lib/Pool.js [hunk 7]
```diff
@@ -102,15 +110,15 @@ class Pool extends EventEmitter {
 
     /**
      * Infinitely looping iterator over available object
-     * @type {DLLArrayIterator}
+     * @type {DequeIterator}
```
Comment (geovanisouza92) [inline lib/Pool.js]: Outdated type

File: lib/Pool.js [hunk 8]
```diff
@@ -184,7 +192,7 @@ class Pool extends EventEmitter {
 
     const pooledResource = this._availableObjects.shift();
     this._dispatchPooledResourceToNextWaitingClient(pooledResource);
-    return;
+    return false;
```
Comment (geovanisouza92) [inline lib/Pool.js]: Seems to be "buggy"

File: lib/Pool.js [hunk 9]
```diff
@@ -373,6 +381,7 @@ class Pool extends EventEmitter {
   _scheduleEvictorRun() {
     // Start eviction if set
     if (this._config.evictionRunIntervalMillis > 0) {
+      // @ts-ignore
       this._scheduledEviction = setTimeout(() => {
```
Comment (geovanisouza92) [inline lib/Pool.js]: Some things I just marked for TS ignore, just to avoid more complex type annotations.

File: lib/Pool.js [hunk 10]
```diff
@@ -402,13 +413,7 @@ class Pool extends EventEmitter {
    * when a new resource is available, passing the resource to the callback.
    * TODO: should we add a seperate "acquireWithPriority" function
    *
-   * @param {Function} callback
-   *   Callback function to be called after the acquire is successful.
-   *   If there is an error preventing the acquisition of resource, an error will
-   *   be the first parameter, else it will be null.
-   *   The acquired resource will be the second parameter.
-   *
```
Comment (geovanisouza92) [inline lib/Pool.js]: I think that this `callback` parameter is really old, right?

File: lib/PoolOptions.js [hunk 11]
```diff
@@ -4,40 +4,45 @@ const PoolDefaults = require("./PoolDefaults");
 
 class PoolOptions {
   /**
-   * @param {Object} config
+   * @param {Object} opts
```
Comment (geovanisouza92) [inline lib/PoolOptions.js]: From here I fixed name mismatches, missing parameters and defaults and updated doc reference.

File: lib/PriorityQueue.js [hunk 12]
```diff
@@ -9,6 +9,7 @@ const Queue = require("./Queue");
 class PriorityQueue {
   constructor(size) {
     this._size = Math.max(+size | 0, 1);
+    /** @type {Queue[]} */
```
Comment (geovanisouza92) [inline lib/PriorityQueue.js]: This solved a lot error messages in this file.

(TS can guess many things with small annotations here and there)

**Comment by sandfox** [comment]
Thanks so much for all this, the original docblocks were just what I put together whilst doing the rewrite to help at the time, but I never actually used them for anything.
The current state of the project is more of an evolution than designed from scratch effort. I'm not against using TS or Flow, but I'd rather do it with the next iteration of the library which had a well thought out api. 

I'd be up for a adding some easy way to test/check the docblocks work with TS, do you have any suggestions?

**Comment by geovanisouza92** [comment]
@sandbox this PR adds type checking with TS without rewriting. It's why it have a tsconfig and TS installed.

**Pull request closed without merging**